homeassistant/components/sky_hub/device_tracker.py

```diff
@@ -31,72 +37,134 @@ class SkyHubDeviceScanner(DeviceScanner):
 
     def __init__(self, config):
         """Initialise the scanner."""
-        _LOGGER.info("Initialising Sky Hub")
+        self._connection_failed = False
+        self._dataparse_failed = False
+        self._log_message("Initialising Sky Hub", level=_INFO)
         self.host = config.get(CONF_HOST, "192.168.1.254")
         self.last_results = {}
         self.url = f"http://{self.host}/"
+        self.success_init = False
 
-        # Test the router is accessible
-        data = _get_skyhub_data(self.url)
+    async def async_connect(self, hass):
+        """Test the router is accessible."""
+        data = await self._async_get_skyhub_data(self.url, hass)
         self.success_init = data is not None
 
-    def scan_devices(self):
+    async def async_scan_devices(self):
         """Scan for new devices and return a list with found device IDs."""
-        self._update_info()
+        await self._async_update_info()
 
         return (device for device in self.last_results)
 
-    def get_device_name(self, device):
+    async def async_get_device_name(self, device):
         """Return the name of the given device or None if we don't know."""
         # If not initialised and not already scanned and not found.
         if device not in self.last_results:
-            self._update_info()
+            await self._async_update_info()
 
             if not self.last_results:
                 return None
 
         return self.last_results.get(device)
 
-    def _update_info(self):
+    async def _async_update_info(self):
         """Ensure the information from the Sky Hub is up to date.
 
         Return boolean if scanning successful.
         """
         if not self.success_init:
             return False
 
-        _LOGGER.info("Scanning")
+        self._log_message("Scanning", level=_INFO)
 
-        data = _get_skyhub_data(self.url)
+        data = await self._async_get_skyhub_data(self.url)
 
         if not data:
-            _LOGGER.warning("Error scanning devices")
             return False
 
         self.last_results = data
 
         return True
 
-
-def _get_skyhub_data(url):
-    """Retrieve data from Sky Hub and return parsed result."""
-    try:
-        response = requests.get(url, timeout=5)
-    except requests.exceptions.Timeout:
-        _LOGGER.exception("Connection to the router timed out")
+    async def _async_get_skyhub_data(self, url, hass=None):
+        """Retrieve data from Sky Hub and return parsed result."""
+        websession = async_get_clientsession(hass if hass else self.hass)
+        parseddata = None
+        try:
+            async with getattr(websession, "get")(url,) as response:
```

Comment:
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices

Reply:
I've not changed the structure of this other than to change to async and remove the exceptions. I'm happy to leave it to the original developer to re-write if needed. If I have time in the future I'll come back and look at it again.

Reply:
We won't accept changes to this part of the integration until the code is extracted.

Reply:
I’m done with that. As I said I’ll come back and move the single rest call external at some point in the future when I have time. Feel free to close the PR and I’ll resubmit when it is done unless someone gets to it before me.

Reply:
If you close it you can reopen later if you wish with history intact.